skyportal/handlers/api/annotation.py

```diff
@@ -176,40 +270,87 @@ def put(self, annotation_id):
-            annotation_id, self.current_user, mode="update", raise_if_none=True
-        )
 
+        try:
+            annotation_id = int(annotation_id)
+        except (TypeError, ValueError):
+            return self.error("Must provide a valid (scalar integer) annotation ID. ")
+
+        if associated_resource_type.lower() == "sources":
+            schema = Annotation.__schema__()
+            try:
+                a = Annotation.get_if_accessible_by(
+                    annotation_id, self.current_user, mode="update", raise_if_none=True
+                )
+            except AccessError:
+                return self.error('Could not find any accessible annotations.')
+            annotation_resource_id_str = str(a.obj_id)
+
+        # add more options using elif
+        else:
+            return self.error(
+                f'Unsupported input "{associated_resource_type}" given as "associated_resource_type" argument.'
+            )
         data = self.get_json()
         group_ids = data.pop("group_ids", None)
         data['id'] = annotation_id
 
-        schema = Annotation.__schema__()
         try:
             schema.load(data, partial=True)
         except ValidationError as e:
             return self.error(f'Invalid/missing parameters: {e.normalized_messages()}')
 
         if group_ids is not None:
-            groups = Group.get_if_accessible_by(
-                group_ids, self.current_user, raise_if_none=True
-            )
+            try:
+                groups = Group.get_if_accessible_by(
+                    group_ids, self.current_user, raise_if_none=True
+                )
+            except AccessError:
+                return self.error('Could not find any accessible groups.')
             a.groups = groups
 
+        if annotation_resource_id_str != resource_id:
+            return self.error(
+                f'Annotation resource ID does not match resource ID given in path ({resource_id})'
+            )
+
         self.verify_and_commit()
-        self.push_all(
-            action='skyportal/REFRESH_SOURCE', payload={'obj_key': a.obj.internal_key}
-        )
+
+        if a.obj.id:  # comment on object, or object related resources
+            self.push_all(
+                action='skyportal/REFRESH_SOURCE',
+                payload={'obj_key': a.obj.internal_key},
+            )
+
         return self.success()
 
     @permissions(['Annotate'])
-    def delete(self, annotation_id):
+    def delete(self, associated_resource_type, resource_id, annotation_id):
         """
         ---
         description: Delete an annotation
         tags:
           - annotations
         parameters:
+          - in: path
+            name: associated_resource_type
+            required: true
+            schema:
+              type: string
+            description: |
+               What underlying data the annotation is on:
+               "sources", or a "spectrum" (this option is not yet supported).
```

Comment:
Related to our last discussion: either implement spectrum annotations here, or move this docs change to another PR.